src/webgpu/idl/javascript.spec.ts

```diff
@@ -0,0 +1,339 @@
+export const description = `
+Validate that various GPU objects behave correctly in JavaScript.
+
+Examples:
+  - GPUTexture, GPUBuffer, GPUQuerySet, GPUDevice, GPUAdapter, GPUAdapter.limits, GPUDevice.limits
+    - return nothing for Object.keys()
+    - adds no properties for {...object}
+    - iterate over keys for (key in object)
+    - throws for (key of object)
+    - return nothing from Object.getOwnPropertyDescriptors
+ - GPUAdapter.limits, GPUDevice.limits
+     - can not be passed to requestAdapter as requiredLimits
+ - GPUAdapter.features, GPUDevice.features
+    - do spread to array
+    - can be copied to set
+    - can be passed to requestAdapter as requiredFeatures
+`;
+
+import { TestCaseRecorder, TestParams } from '../../common/framework/fixture.js';
+import { makeTestGroup } from '../../common/framework/test_group.js';
+import { keysOf } from '../../common/util/data_tables.js';
+import { getGPU } from '../../common/util/navigator_gpu.js';
+import { assert, objectEquals, unreachable } from '../../common/util/util.js';
+import { getDefaultLimitsForAdapter, kLimits } from '../capability_info.js';
+import {
+  DeviceSelectionDescriptor,
+  GPUTest,
+  GPUTestSubcaseBatchState,
+  initUncanonicalizedDeviceDescriptor,
+} from '../gpu_test.js';
+import { CanonicalDeviceDescriptor, DescriptorModifier } from '../util/device_pool.js';
+
+function addAllFeatures(adapter: GPUAdapter, desc: CanonicalDeviceDescriptor | undefined) {
+  const descWithMaxLimits: CanonicalDeviceDescriptor = {
+    defaultQueue: {},
+    ...desc,
+    requiredFeatures: [...adapter.features] as GPUFeatureName[],
+    requiredLimits: {},
```

Comment:
nit: function says "addAllFeatures" but I think it also removes all limit requests.

Reply:
fixed